### ISSUE TITLE
[clang][static analyzer]alpha.deadcode.UnreachableCode false positive on cxxTryStmt

### ISSUE THREAD

**Issue by mzyKi** [llvm/llvm-project#90162]
```cpp
  class BaseException {};
  
  class DerivedException : public BaseException {};
  
  void foo();
  
  void f4() {
    try {
      foo();
    } catch (BaseException &b) {
    } catch (DerivedException &d) {
    }
  }
```

When I run ```clang-tidy --allow-enabling-analyzer-alpha-checkers -checks=-*,clang-analyzer-alpha.deadcode.UnreachableCode test.cpp```, the result like as following:

```bash
test.cpp:8:3: warning: This statement is never executed [clang-analyzer-alpha.deadcode.UnreachableCode]
    8 |   try {
      |   ^~~~~
    9 |     foo();
      |     ~~~~~~
   10 |   } catch (BaseException &b) {
      |   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   11 |   } catch (DerivedException &d) {
      |   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   12 |   }
      |   ~
test.cpp:8:3: note: This statement is never executed
    8 |   try {
      |   ^~~~~
    9 |     foo();
      |     ~~~~~~
   10 |   } catch (BaseException &b) {
      |   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   11 |   } catch (DerivedException &d) {
      |   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   12 |   }
      |   ~

```

I think this is false positive. If not , I am not clear about reason.

**Comment by steakhal** [comment]
I remember this. This is easy to fix, just have an early return at place where we list all the unsupported Stmts.

**Comment by lolloz98** [comment]
I will give this issue a try, I am a newby on llvm codebase so I might have some questions (please assign it to me, because on a previous issue I just commented and after a couple of days another guy made PR while I was waiting to test my changes......)
Thank you :)

**Comment by steakhal** [comment]
Personally, I prefer not assigning issues to people because I fear getting issues stale, and other contributors could be discouraged to attempt to fix an issue even if it was assigned to someone else for a long time ago.
Consequently, from my perspective first comes first served for PRs. But I also get your perspective.

**Comment by lolloz98** [comment]
Completely understand your point of view

**Comment by lolloz98** [comment]
> clang-tidy --allow-enabling-analyzer-alpha-checkers -checks=-*,clang-analyzer-alpha.deadcode.UnreachableCode

Sorry, but I have been debugging this issue and I was wondering if I actually understand what the supposed output should be. I think that for this code it should be `f4 is unreachable` because f4 is not called anywhere.

To support this hypothesis I have tried to run the same command on:
```
  class BaseException {};
  
  class DerivedException : public BaseException {};

  void foo();
  
  void f4() {
    try {
      foo();
    } catch (BaseException &b) {
    } catch (DerivedException &d) {
    }
  }
  int main() {
    f4();
    return 0;
  }
```

does not generate the warning. Moreover, I am not finding a place where there are unsupported statements so if you could expand on that, it would be very helpful :)

**Comment by steakhal** [comment]
This checker supposed to diagnose the first unreachable statement of a control-flow block.
By only looking at a single translation unit, we cannot know if a different translation unit actually calls this external-linkage function. Consequently, we don't report such things.

I can see that you are really interested in this checker.
I can tell you that, this "cxxTryStmt" issue is not the only thing hindering this checker.
I have actually a list of things, some of which is difficult to tackle.
I also have a private branch where I played with fixing these, but I have to re-measure the reports at scale, because it appeared to me that it suppressed almost all the issues - including possibly valid reports.
I'll come back to this, in a few days, and maybe even repeat my experiment to double-check if I just messed it up.

**Comment by lolloz98** [comment]
Interesting.

After your comment I have debugged a bit more and I have found out this call https://github.com/llvm/llvm-project/blob/85f28cf431911ca505b759c155bb62e8be7151ae/clang/lib/StaticAnalyzer/Checkers/UnreachableCodeChecker.cpp#L113

And the interesting part is that inside it checks if CB has predecessors and in case of the code provided the tryStmt seems not to have one. Anyway I would be glad to learn more :) if you want we can connect via discord (I have the same nickname there) :)

**Comment by steakhal** [comment]
> Interesting.
> 
> After your comment I have debugged a bit more and I have found out this call
> 
> https://github.com/llvm/llvm-project/blob/85f28cf431911ca505b759c155bb62e8be7151ae/clang/lib/StaticAnalyzer/Checkers/UnreachableCodeChecker.cpp#L113

Yes, that place!

> And the interesting part is that inside it checks if CB has predecessors and in case of the code provided the tryStmt seems not to have one.

Exactly, and that's why that's the first basic block in the CFG which is "not visited", thus we report the first Stmt of it as "unreachable statement". But don't take my word for it, I only have vague memories of this checker.

**Comment by lolloz98** [comment]
Yeah, I tried doing a couple of tries, but unfortunately I have been a bit swamped and I will not be able to use computer for a couple of weeks, so I will unassign me from this issue (if no one picks it up I will get it back in a couple of weeks, but no worries if you want to investigate :))

**Comment by Damiencastro** [comment]
Hi, I would like to work on this issue. No need to assign me the issue given your earlier comment.
